oslo/torch/nn/parallel/data_parallel/zero/sharded_optim/_utils.py

```diff
@@ -0,0 +1,284 @@
+    )
+
+
+def flatten(input_: Iterable[torch.Tensor]) -> torch.Tensor:
+    """
+    Flatten the input tensor into a 1D tensor.
+
+    Args:
+        input_ (Iterable[torch.Tensor]): The input tensor to be flattened.
+
+    Returns:
+        torch.Tensor: The flattened tensor.
+    """
+    return _flatten_dense_tensors(input_)
+
+
+def unflatten(flat: torch.Tensor, tensors: Iterable[torch.Tensor]) -> torch.Tensor:
+    """
+    Unflatten the flattened tensor back to its original shape.
+
+    Args:
+        flat (torch.Tensor): The flattened tensor.
+        tensors (Iterable[torch.Tensor]): The original tensor shape.
+
+    Returns:
+        torch.Tensor: The unflattened tensor.
+    """
+    return _unflatten_dense_tensors(flat, tensors)
+
+
+def calculate_global_norm_from_list(norm_list: List[float]) -> float:
+    """
+    Compute the total global norm from a list of norms.
+
+    Args:
+        norm_list (List[float]): List of norms to compute the total global norm from.
+
+    Returns:
+        float: Total global norm.
+    """
+    total_norm = 0.0
+    for norm in norm_list:
+        total_norm += norm**2.0
+    return math.sqrt(total_norm)
+
+
+def reduce_tensor_dp_group(
+    tensor: torch.Tensor,
+    dtype: Optional[torch.dtype] = None,
+    dst_local_rank: Optional[int] = None,
+    dst_global_rank: Optional[int] = None,
+    group: Optional[dist.ProcessGroup] = None,
+) -> torch.Tensor:
+    """
+    Reduce a tensor in the data parallel process group.
+
+    Args:
+        tensor (torch.Tensor): Tensor to reduce/all-reduce.
+        dtype (Optional[torch.dtype]): Data type used in communication.
+        dst_local_rank (Optional[int]): Local rank of the destination node.
+        dst_global_rank (Optional[int]): Global rank of the destination node.
+        group (Optional[dist.ProcessGroup]): Process group for the reduction.
+
+    Returns:
+        tensor (torch.Tensor): The reduced tensor.
+    """
+
+    # use the original dtype
+    if dtype is None:
+        dtype = tensor.dtype
+
+    # cast the data to specified dtype for reduce/all-reduce
+    if tensor.dtype != dtype:
+        tensor_to_reduce = tensor.to(dtype)
+    else:
+        tensor_to_reduce = tensor
+
+    world_size = dist.get_world_size(group=group)
+    tensor_to_reduce.div_(world_size)
+
+    # if rank is None, all reduce will be used
+    # else, reduce is used
+    use_all_reduce = dst_local_rank is None
+
+    if use_all_reduce:
+        dist.all_reduce(tensor_to_reduce, group=group)
+    else:
+        dist.reduce(tensor=tensor_to_reduce, dst=dst_global_rank, group=group)
+
+    # recover the original dtype
+    if tensor.dtype != dtype and tensor is not tensor_to_reduce:
+        local_rank = dist.get_rank(group=group)
+        if use_all_reduce or dst_local_rank == local_rank:
+            tensor.copy_(tensor_to_reduce)
+
+    return tensor
+
+
+def has_inf_or_nan(tensor: torch.Tensor) -> bool:
+    """
```

Comment:
How about using `torch.isfinite`?

Reply:
I'm not the author of the original code, so I don't know the exact intention, but when I searched, it seems that there is some kind of floating point exception related to fp16. The intention of this function is probably to have a bigger role in checking if an error occurs when converting fp16 to fp32 and merging with the accumulated gradient (rather than simply checking inf which is nan). I tried to reproduce it, but had difficulties. I'm so sorry. Thanks for the review!

Reply:
I think using the original function is better. Because there are some side effects.

Reply:
As suggested by you, I used Torch's original function (`torch.isinf` and `torch.isnan`).

Reply:
Then, you may be able to use `torch.isfinite`. in line 137.
https://pytorch.org/docs/stable/generated/torch.isfinite.html

Reply:
Your suggestion has been reflected. thank you!

Reply:
The original function I mentioned is ColossalAI's function.

Reply:
I misunderstood your words. Reverted to the original function. thank you.